### PR TITLE
Add nextcloud provider

### PR DESCRIPTION
**Pull request by Ramblurr**
## Description

This PR adds the [Nextcloud](https://nextcloud.com/) provider, that allows for authentication of users on a Nextcloud instance.

## Motivation and Context

For some time now Nextcloud has had an Oauth2 implementation that allows for arbitrary 3rd party tools to act as oauth2 clients and perform authentication for users hosted on the Nextcloud instance.

Docs: https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html

## How Has This Been Tested?

* I've included a test suite
* I'm running this in production on a large Nextcloud instance

## Checklist

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,7 @@
 
 ## Changes since v3.2.0
 
+- [XX]() Add Nextcloud provider
```
Comment (JoelSpeed) [inline CHANGELOG.md]: Please update this

**Comment by Ramblurr** [comment]
Thanks for the review. I'm leaving for vacation soon, so  I'll get back to this in several weeks.

**Comment by Ramblurr** [comment]
I'm back at this again, sorry for the delay.

I rebased and patched my PR given the recent shakeup for 4.0.0, and I've made the requested changes.

The Travis build is failing due to a linting error in code I didn't touch:

```
GO111MODULE=on "/home/travis/gopath/bin/golangci-lint" run
providers/azure_test.go:116:47: string `Bearer imaginary_access_token` has 3 occurrences, make it a constant (goconst)

			} else if r.Header.Get("Authorization") != "Bearer imaginary_access_token" {
			                                           ^
Makefile:20: recipe for target 'lint' failed
make: *** [lint] Error 1
The command "./configure && make test" exited with 2.
```
This is code in `master` already.

If I disable linting and run just the tests, they pass fine locally.

**Comment by steakunderscore** [comment]
>`} else if r.Header.Get("Authorization") != "Bearer imaginary_access_token" {`
This came up somewhere else in the repo. Just change the string so it's unique. Could use something like `azure_imaginary_access_token`. I think this should fix it.

**Comment by Ramblurr** [comment]
Should I add that as a commit to this PR?

**Comment by ploxiln** [comment]
In the function `testNextcloudBackend()` you added the line:

```go
			} else if r.Header.Get("Authorization") != "Bearer imaginary_access_token" {
```

and that is what caused this dumb linter rule to be triggered. So to quiet the rule, you can change your tests to use a different mock access token string.

**Comment by Ramblurr** [comment]
Oh! I see, the linter is comparing the strings across the entire project. Sorry about that. That explains why this isn't present on master :)

Pushed the fix.

File: docs/2_auth.md [hunk 2]
```diff
@@ -271,6 +272,33 @@ In this case, you can set the `-skip-oidc-discovery` option, and supply those re
...
+configuration, environment variable, or command line argument. Depending
+on whether your Nextcloud instance is using pretty urls your urls may be of the
+form `/index.php/apps/oauth2/*` or `/apps/oauth2/*`.
+
+Refer to the [OAuth2
+documentation](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html)
+to setup the client id and client secret. Your "Redirection URI" will be
+`https://internalapp.yourcompany.com/oauth2/callback`.
+
+
```
Comment (syscll) [inline docs/2_auth.md]: Nit: remove new line.

File: providers/nextcloud.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package providers
+
+import (
+	"fmt"
+
```
Comment (syscll) [inline providers/nextcloud.go]: Nit: remove new line.

**Comment by Ramblurr** [comment]
> Added a few tiny nits (sorry).
fixed!

> Are you ok to add yourself to the CODEOWNERS for this provider? Totally optional of course 😄
If you're referring to .github/CODEOWNERS, then I've already added myself there :)

**Comment by syscll** [comment]
This has been open for a while so let's get the changelog conflicts resolved and we can merge.

@JoelSpeed: if you've time to re-review then that would be great!

**Comment by Ramblurr** [comment]
Fixed the CHANGELOG conflict.

File: docs/2_auth.md [hunk 4]
```diff
@@ -288,6 +289,32 @@ In this case, you can set the `-skip-oidc-discovery` option, and supply those re
...
+Refer to the [OAuth2
+documentation](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html)
+to setup the client id and client secret. Your "Redirection URI" will be
+`https://internalapp.yourcompany.com/oauth2/callback`.
+
+```
+    -provider nextcloud
+    -client-id <from nextcloud admin>
+    -client-secret <from nextcloud admin>
+    -login-url="<your gitlab url>/index.php/apps/oauth2/authorize"
```
[the remark this answers is not in the document]
Comment (Ramblurr) [reply docs/2_auth.md]: Woops, you are right. Should definitely be nextcloud.

**Comment by Ramblurr** [comment]
Fixed doc typo and rebased master.

**Pull request merged**